docs/design/speed-test.md

```diff
@@ -0,0 +1,118 @@
+
+This Upload Speed Test CRD will help in such scenarios by proactively measuring the upload speed from the OpenShift cluster to the cloud storage configured in OADP’s BSL. By identifying
+performance bottlenecks ahead of backup or restore operations, administrators can take corrective actions to avoid failures during critical operations.
+
+## Goals
+
+- 	Utilize BackupStorageLocation (BSL) settings from the OADP operator’s DPA CRD for storage configurations and credentials.
+-   Measure upload speeds from the OpenShift cluster to the configured object storage.
+-   Update the test results in the status of the UploadSpeedTest CRD for visibility and tracking.
+-   Ensure smooth integration with OADP operator-managed backup storage location configurations.
+
+
+## Non-Goals
+
+- This design will not create or modify BackupStorageLocation entries in OADP.
+- It will not implement download or latency tests, focusing solely on upload speed.
+- Scheduling of recurring tests is not supported in the initial version.
+
+## High-Level Design
+Components involved and their responsibilities:
+
+- UploadSpeedTest (UST) CRD:
+  - Captures the BSL configuration used for testing directly in the CRD.
+  - Holds the test results, including success/failure and upload speed.
+- UploadSpeedTest (UST) Controller:
+  - Watches for changes to the UploadSpeedTest CRs and executes uploads based on the provided BSL configuration. 
+  - Uses appropriate cloud SDKs (AWS, Azure, or GCP) to perform uploads.
+  - Updates the status of UST CR with upload speed, upload success/failure and error message if any.
+- OADP/DPA-UST Integration:
+  - Fetches BackupStorageLocation (BSL) details from the OADP operator’s DPA CR. 
+  - Populates the UploadSpeedTest CR with relevant configurations from the BSL.
+  - Creates the UST CR for the BSL configured in DPA instance.
+## Detailed Design
+
+Proposed Upload Speed Test CRD:
+
+```yaml
+apiVersion: uploadspeedtest.io/v1alpha1
+kind: UploadSpeedTest
+metadata:
+  name: upload-speed-test-sample
+spec:
+  dpaNamespace: oadp-operator    # Namespace of the DPA CR
+  bslConfig:                     # Configuration from BackupStorageLocation
+    name: aws-backup-location
+    provider: aws
+    region: us-east-1
+    bucket: my-backup-bucket
+    uploadSpeedTest:
+      fileSize: 10MB             # Test file size
+      testTimeout: 60s           # Timeout for upload test
+  cloudProviderSecretRef: aws-credentials   # Secret for authentication
+status:
+  lastTested: "2024-10-08T10:00:00Z"
+  uploadSpeedMbps: 55.3
+  uploadSuccess: true
+  errorMessage: ""
+```
+
+Changes to DPA CRD BSL Configuration spec:
+
+```yaml
+apiVersion: oadp.openshift.io/v1alpha1
+kind: DataProtectionApplication
+metadata:
+  name: sample-dpa
+spec:
+  backupStorageLocations:
+    - name: aws-backup-location
+      provider: aws
+      config:
+        region: us-east-1
+        bucket: my-backup-bucket
+        uploadSpeedTest:
```

Comment:
If we're not going to support automatic speed checks, I'm confused why the DPA needs to change initially.

Reply:
Auto-Check or not depends on the user configuration here. For each BSL configured we have a `uploadSpeedTest` struct that has `enabled` flag and `uploadSpeedTest` config. If the user sets these then UST CR would be automatically created by OADP Operator for that particular BSL, else the user could create the UST CR themselves.